### PR TITLE
Add Carbon Intensity Average

### PR DESCRIPTION
**Pull request by juzuluag**
Issue Number: 169

## Summary
Add Carbon intensity average calculation based on location and dates.
## Changes

- Enhance CarbonAwareAggregator to calculate EmissionsData Rating average.

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [x] Added any new Tests?
- [ ] Documentation Updates Made?
- [x] This is not a breaking change. If it is, please describe it below.

## Is this a breaking change?
N/A

## Anything else?
Might require some small rework after JSON rework is in place.

### CODE REVIEW

File: src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs [hunk 1]
```diff
@@ -3,24 +3,41 @@
...
-        public CarbonAwareAggregator(ILogger<CarbonAwareAggregator> logger, ICarbonAware plugin)
+    private void ValidateAverageProps(IDictionary props)
+    {
+        if (!props.Contains(CarbonAwareConstants.Locations) ||
+            !props.Contains(CarbonAwareConstants.Start) ||
+            !props.Contains(CarbonAwareConstants.End))
         {
-            this._logger = logger ?? throw new ArgumentNullException(nameof(logger));
-            this._plugin = plugin;
+            throw new ArgumentException("Missing properties to calculate average");
```
Comment (JenMadiedo) [inline src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:35]: 🐑 in the future we might want to split this out so that we can explicitly tell users which props they're missing

Comment (juzuluag) [reply src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:35]: Agree. I'll wait till #26 lands to make other changes that affects this.

File: src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs [hunk 2]
```diff
@@ -3,66 +3,83 @@
...
-            if (!DateTimeOffset.TryParse(dto.ToString(), out defaultDto))
-            {
-                Exception ex = new ArgumentException("Failed to parse" + field + "field. Must be a valid DateTimeOffset");
-                _logger.LogError("argument exception", ex);
-                throw ex;
-            }
+        return defaultDto;
+    }
 
-            return defaultDto;
```
Comment (pritipath) [inline src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:52]: The method name - GetOrDefautlLocation is misleading. We are not setting a default Location if not present but throwing an Exception. Perhaps change it to GetLocationOrThrow?

Comment (JenMadiedo) [reply src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:52]: Whoops, that's coming from my PR that he's pulling in, I'll fix upstream

**Review by JenMadiedo — APPROVED** [review]
LGTM

File: src/dotnet/CarbonAware.Aggregators/test/CarbonAwareAggregatorTests.cs [hunk 3]
```diff
@@ -1,13 +1,132 @@
...
+        var average = await aggregator.CalcEmissionsAverageAsync(props);
+        Assert.GreaterOrEqual(average, expected);
+    }
+
+    [TestCase(null, null, null)]
+    [TestCase("westus", null, null)]
+    [TestCase(null, "2021-12-19", null)]
+    [TestCase(null, null, "2021-12-20")]
+    [TestCase(null, "2021-12-19", "2021-12-20")]
+    public void Test_Emissions_Average_Missing_Properties(string location, string startTime, string endTime)
```
Comment (JenMadiedo) [inline src/dotnet/CarbonAware.Aggregators/test/CarbonAwareAggregatorTests.cs:70]: You'll need to change this test I think. Because we do the defaulting in the aggregator, not sure what the expected behavior is here now

File: src/dotnet/CarbonAware.DataSources.WattTime/src/Configuration/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
 using CarbonAware.Interfaces;
+using CarbonAware.Tools.WattTimeClient.Configuration;
+
 
 namespace CarbonAware.DataSources.WattTime.Configuration;
 
 public static class ServiceCollectionExtensions
 {
     public static void AddWattTimeDataSourceService(this IServiceCollection services)
     {
```
Comment (pritipath) [inline src/dotnet/CarbonAware.DataSources.WattTime/src/Configuration/ServiceCollectionExtensions.cs:13]: Nit pick and I am also unaware with the coding conventions in C#, so ignore if this seems irrelevant. Do we need to create vars or can the code block be simplified by reducing it to a line -
services.ConfigureWattTimeClient(new ConfigurationBuilder().Build());

Comment (juzuluag) [reply src/dotnet/CarbonAware.DataSources.WattTime/src/Configuration/ServiceCollectionExtensions.cs:13]: Yup, it can be done in oneline too. I'll keep this way since it requires potentially to set env vars and/or config file.

File: src/dotnet/CarbonAware.Aggregators/test/CarbonAwareAggregatorTests.cs (hunk 3 above)
Comment (juzuluag) [reply src/dotnet/CarbonAware.Aggregators/test/CarbonAwareAggregatorTests.cs:70]: Hi @JenMadiedo . Not sure I get it. You mean "defaulting" in the aggregator what specific ?

File: .gitignore [hunk 5]
```diff
@@ -16,4 +16,5 @@ TestResults
 !.vscode/
 .vscode/*
 !.vscode/launch.json
-!.vscode/tasks.json
\ No newline at end of file
+!.vscode/tasks.json
```
Comment (gfmatthews) [inline .gitignore:19]: So are we keeping this or not?

File: src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs [hunk 6]
```diff
@@ -3,59 +3,78 @@
...
-        private readonly ILogger<CarbonAwareAggregator> _logger;
+        this._logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        this._dataSource = dataSource;
+    }
 
-        private readonly ICarbonIntensityDataSource _dataSource;
+    public async Task<IEnumerable<EmissionsData>> GetEmissionsDataAsync(IDictionary props)
+    {
+        DateTimeOffset end = GetOffsetOrDefault(props, CarbonAwareConstants.End, DateTimeOffset.Now);
+        DateTimeOffset start = GetOffsetOrDefault(props, CarbonAwareConstants.Start, end.AddDays(-7));
```
Comment (gfmatthews) [inline src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:23]: props for using DateTime objects :)

File: .gitignore (hunk 5 above)
Comment (juzuluag) [reply .gitignore:19]: I think so. I'm just adding an CR\LF to add the next entry (.DS_Store)

File: src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs [hunk 7]
```diff
@@ -3,59 +3,78 @@
...
-                return locations;
-            }
-            Exception ex = new ArgumentException("locations parameter must be provided and be non empty");
-            _logger.LogError("argument exception", ex);
-            throw ex;
+    public async Task<double> CalcEmissionsAverageAsync(IDictionary props)
+    {
+        ValidateAverageProps(props);
+        var list = await GetEmissionsDataAsync(props);
+        var value = list.Any() ? list.Select(x => x.Rating).Average() : 0;
```
Comment (gfmatthews) [inline src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:65]: Optional: Super readable but might be worth a comment explaining what you're doing here due to use of syntactic sugar.

I also respect the idea that code is self-documenting and so if you think this is excessive, I'm for that as well

**Review by gfmatthews — APPROVED** [review]
Looks great y'all!

File: src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs (hunk 7 above)
Comment (juzuluag) [reply src/dotnet/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs:65]: Sure, I'll add a comment. Thx
